### ISSUE TITLE
zmq4: data race in socket.Close

### ISSUE THREAD

**Issue by sbinet** [go-zeromq/zmq4#50]
it would seem #47 introduced a data race on `socket.closedConns`.

```
==================
WARNING: DATA RACE
Write at 0x00c000278130 by goroutine 87:
  runtime.closechan()
      /home/travis/.gimme/versions/go1.11.13.linux.amd64/src/runtime/chan.go:327 +0x0
  github.com/go-zeromq/zmq4.(*socket).Close.func1()
      /home/travis/gopath/src/github.com/go-zeromq/zmq4/socket.go:92 +0x6e
  github.com/go-zeromq/zmq4.(*socket).Close()
      /home/travis/gopath/src/github.com/go-zeromq/zmq4/socket.go:116 +0x317
  github.com/go-zeromq/zmq4.(*dealerSocket).Close()
      /home/travis/gopath/src/github.com/go-zeromq/zmq4/dealer.go:26 +0x67
  runtime.call32()
      /home/travis/.gimme/versions/go1.11.13.linux.amd64/src/runtime/asm_amd64.s:522 +0x3a
  testing.(*common).Fatalf()
      /home/travis/.gimme/versions/go1.11.13.linux.amd64/src/testing/testing.go:634 +0x94
  github.com/go-zeromq/zmq4_test.TestRouterDealer.func1()
      /home/travis/gopath/src/github.com/go-zeromq/zmq4/zmq4_routerdealer_test.go:233 +0xf09
  testing.tRunner()
      /home/travis/.gimme/versions/go1.11.13.linux.amd64/src/testing/testing.go:827 +0x162

Previous read at 0x00c000278130 by goroutine 50:
  runtime.chansend()
      /home/travis/.gimme/versions/go1.11.13.linux.amd64/src/runtime/chan.go:140 +0x0
  github.com/go-zeromq/zmq4.(*socket).scheduleRmConn()
      /home/travis/gopath/src/github.com/go-zeromq/zmq4/socket.go:290 +0x105
  github.com/go-zeromq/zmq4.(*socket).scheduleRmConn-fm()
      /home/travis/gopath/src/github.com/go-zeromq/zmq4/socket.go:184 +0x4b
  github.com/go-zeromq/zmq4.(*Conn).notifyOnCloseError()
      /home/travis/gopath/src/github.com/go-zeromq/zmq4/conn.go:443 +0xac
  github.com/go-zeromq/zmq4.(*Conn).SetClosed()
      /home/travis/gopath/src/github.com/go-zeromq/zmq4/conn.go:420 +0xab
  github.com/go-zeromq/zmq4.(*Conn).checkIO()
      /home/travis/gopath/src/github.com/go-zeromq/zmq4/conn.go:435 +0x150
  github.com/go-zeromq/zmq4.(*Conn).read()
      /home/travis/gopath/src/github.com/go-zeromq/zmq4/conn.go:333 +0xcf7
  github.com/go-zeromq/zmq4.(*qreader).listen()
      /home/travis/gopath/src/github.com/go-zeromq/zmq4/msgio.go:103 +0xe7

Goroutine 87 (running) created at:
  testing.(*T).Run()
      /home/travis/.gimme/versions/go1.11.13.linux.amd64/src/testing/testing.go:878 +0x659
  github.com/go-zeromq/zmq4_test.TestRouterDealer()
      /home/travis/gopath/src/github.com/go-zeromq/zmq4/zmq4_routerdealer_test.go:96 +0x2b2
  testing.tRunner()
      /home/travis/.gimme/versions/go1.11.13.linux.amd64/src/testing/testing.go:827 +0x162

Goroutine 50 (running) created at:
  github.com/go-zeromq/zmq4.(*qreader).addConn()
      /home/travis/gopath/src/github.com/go-zeromq/zmq4/msgio.go:66 +0xa2
  github.com/go-zeromq/zmq4.(*socket).addConn()
      /home/travis/gopath/src/github.com/go-zeromq/zmq4/socket.go:255 +0x40d
  github.com/go-zeromq/zmq4.(*socket).Dial()
      /home/travis/gopath/src/github.com/go-zeromq/zmq4/socket.go:241 +0x73d
  github.com/go-zeromq/zmq4.(*dealerSocket).Dial()
      /home/travis/gopath/src/github.com/go-zeromq/zmq4/dealer.go:47 +0x7b
  github.com/go-zeromq/zmq4_test.TestRouterDealer.func1.2.1()
      /home/travis/gopath/src/github.com/go-zeromq/zmq4/zmq4_routerdealer_test.go:182 +0xc0
  golang.org/x/sync/errgroup.(*Group).Go.func1()
      /home/travis/gopath/pkg/mod/golang.org/x/sync@v0.0.0-20190423024810-112230192c58/errgroup/errgroup.go:57 +0x64
==================
panic: send on closed channel

goroutine 494 [running]:
github.com/go-zeromq/zmq4.(*socket).scheduleRmConn(0xc000162500, 0xc0000e7220)
	/home/travis/gopath/src/github.com/go-zeromq/zmq4/socket.go:290 +0x106
github.com/go-zeromq/zmq4.(*socket).scheduleRmConn-fm(0xc0000e7220)
	/home/travis/gopath/src/github.com/go-zeromq/zmq4/socket.go:184 +0x4c
github.com/go-zeromq/zmq4.(*Conn).notifyOnCloseError(0xc0000e7220)
	/home/travis/gopath/src/github.com/go-zeromq/zmq4/conn.go:443 +0xad
github.com/go-zeromq/zmq4.(*Conn).SetClosed(0xc0000e7220)
	/home/travis/gopath/src/github.com/go-zeromq/zmq4/conn.go:420 +0xac
github.com/go-zeromq/zmq4.(*Conn).checkIO(0xc0000e7220, 0x6fbac0, 0xc000247a40)
	/home/travis/gopath/src/github.com/go-zeromq/zmq4/conn.go:435 +0x151
github.com/go-zeromq/zmq4.(*Conn).read(0xc0000e7220, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0)
	/home/travis/gopath/src/github.com/go-zeromq/zmq4/conn.go:333 +0xcf8
github.com/go-zeromq/zmq4.(*qreader).listen(0xc000213b30, 0x6fcac0, 0xc000262c40, 0xc0000e7220)
	/home/travis/gopath/src/github.com/go-zeromq/zmq4/msgio.go:103 +0xe8
created by github.com/go-zeromq/zmq4.(*qreader).addConn
	/home/travis/gopath/src/github.com/go-zeromq/zmq4/msgio.go:66 +0xa3
FAIL	github.com/go-zeromq/zmq4	19.152s
travis_time:end:0610ee78:start=1571132857856874905,finish=1571132886612396666,duration=28755521761,event=script
```

**Comment by Inphi** [comment]
taking a look...

**Comment by sbinet** [comment]
fixed by #53.
